File: examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/llm/LLMService.kt
package com.runanywhere.runanywhereai.llm

import kotlinx.coroutines.flow.Flow

/**
 * Base interface for all LLM service implementations
 * 
 * This interface provides a unified API for interacting with different LLM frameworks
 * including MediaPipe, ONNX Runtime, TensorFlow Lite, llama.cpp, ExecuTorch, MLC-LLM, 
 * Gemini Nano, Android AI Core, and picoLLM.
 * 
 * All implementations should handle model loading, text generation, and resource cleanup
 * in a consistent manner while abstracting away framework-specific details.
 * 
 * @sample Basic usage:
 * ```kotlin
 * val service = MediaPipeService(context)
 * service.initialize("/path/to/model.bin")
 * val result = service.generate("Hello, world!", GenerationOptions())
 * println("Generated: ${result.text}")
 * service.release()
 * ```
 */
interface LLMService {
    val name: String
    val isInitialized: Boolean
    
    /**
     * Initialize the LLM service with a model
     * 
     * @param modelPath Path to the model file (format depends on framework)
     * @throws IllegalArgumentException if model path is invalid
     * @throws IllegalStateException if initialization fails
     */
    suspend fun initialize(modelPath: String)
    
    /**
     * Generate text based on the prompt
     * 
     * @param prompt Input text to generate from
     * @param options Generation parameters (temperature, max tokens, etc.)
     * @return GenerationResult containing generated text and performance metrics
     * @throws IllegalStateException if service is not initialized
     */
    suspend fun generate(prompt: String, options: GenerationOptions = GenerationOptions()): GenerationResult
    
    /**
     * Stream generation results token by token
     * 
     * This method provides real-time streaming of generated text, useful for
     * displaying incremental results to users as they are generated.
     * 
     * @param prompt Input text to generate from
     * @param options Generation parameters
     * @return Flow of GenerationResult objects with partial text and cumulative metrics
     * @throws IllegalStateException if service is not initialized
     */
    fun generateStream(prompt: String, options: GenerationOptions = GenerationOptions()): Flow<GenerationResult>
    
    /**
     * Get information about the loaded model
     */
    fun getModelInfo(): ModelInfo?
    
    /**
     * Release resources
     */
    suspend fun release()
}

/**
 * Options for text generation
 */
data class GenerationOptions(
    val maxTokens: Int = 150,
    val temperature: Float = 0.7f,
    val topP: Float = 0.9f,
    val topK: Int = 40,
    val repetitionPenalty: Float = 1.1f,
    val stopSequences: List<String> = emptyList(),
    val presencePenalty: Float? = null,
    val frequencyPenalty: Float? = null
)

/**
 * Information about the loaded model
 */
data class ModelInfo(
    val name: String,
    val sizeBytes: Long,
    val parameters: Long? = null,
    val quantization: String? = null,
    val format: String,
    val framework: LLMFramework
)

/**
 * Supported LLM frameworks
 */
enum class LLMFramework {
    MEDIAPIPE,
    ONNX_RUNTIME,
    TFLITE,
    LLAMA_CPP,
    EXECUTORCH,
    MLC_LLM,
    GEMINI_NANO,
    PICOLLM,
    AI_CORE
}

/**
 * Chat message for conversation
 */
data class ChatMessage(
    val role: ChatRole,
    val content: String,
    val timestamp: Long = System.currentTimeMillis()
)

/**
 * Chat roles
 */
enum class ChatRole {
    USER, ASSISTANT, SYSTEM
}

/**
 * Reason why generation finished
 */
enum class FinishReason {
    COMPLETED,
    MAX_TOKENS,
    STOP_SEQUENCE,
    ERROR,
    CANCELLED
}

/**
 * Performance metrics for generation
 */
data class PerformanceMetrics(
    val totalTime: Long,
    val timeToFirstToken: Long,
    val tokensPerSecond: Float,
    val tokenCount: Int,
    val memoryUsed: Long,
    val cpuUsage: Float? = null,
    val batteryTemperature: Float? = null
)

File: sdk/runanywhere-swift/Sources/RunAnywhere/Protocols/LLMService.swift
import Foundation

/// Protocol for LLM service implementations
public protocol LLMService: AnyObject {
    /// Initialize the service with a model
    /// - Parameter modelPath: Path to the model file
    func initialize(modelPath: String) async throws

    /// Generate text from a prompt
    /// - Parameters:
    ///   - prompt: The input prompt
    ///   - options: Generation options
    /// - Returns: Generated text
    func generate(prompt: String, options: GenerationOptions) async throws -> String

    /// Stream generate text from a prompt
    /// - Parameters:
    ///   - prompt: The input prompt
    ///   - options: Generation options
    ///   - onToken: Callback for each generated token
    func streamGenerate(
        prompt: String,
        options: GenerationOptions,
        onToken: @escaping (String) -> Void
    ) async throws

    /// Clean up resources
    func cleanup() async

    /// Get current memory usage
    /// - Returns: Memory usage in bytes
    func getModelMemoryUsage() async throws -> Int64

    /// Check if the service is ready
    var isReady: Bool { get }

    /// Get model information
    var modelInfo: LoadedModelInfo? { get }

    /// Set generation context
    /// - Parameter context: The context to use
    func setContext(_ context: Context) async

    /// Clear generation context
    func clearContext() async
}

/// Information about a loaded model
public struct LoadedModelInfo {
    public let id: String
    public let name: String
    public let framework: LLMFramework
    public let format: ModelFormat
    public let memoryUsage: Int64
    public let contextLength: Int
    public let loadedAt: Date
    public let configuration: HardwareConfiguration

    public init(
        id: String,
        name: String,
        framework: LLMFramework,
        format: ModelFormat,
        memoryUsage: Int64,
        contextLength: Int,
        loadedAt: Date = Date(),
        configuration: HardwareConfiguration
    ) {
        self.id = id
        self.name = name
        self.framework = framework
        self.format = format
        self.memoryUsage = memoryUsage
        self.contextLength = contextLength
        self.loadedAt = loadedAt
        self.configuration = configuration
    }
}


/// LLM service errors
public enum LLMServiceError: LocalizedError {
    case notInitialized
    case modelNotLoaded
    case generationFailed(String)
    case contextTooLong(Int, Int)
    case unsupportedOperation(String)
    case frameworkError(FrameworkError)

    public var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Service not initialized"
        case .modelNotLoaded:
            return "Model not loaded"
        case .generationFailed(let reason):
            return "Generation failed: \(reason)"
        case .contextTooLong(let current, let max):
            return "Context too long: \(current) tokens (max: \(max))"
        case .unsupportedOperation(let operation):
            return "Unsupported operation: \(operation)"
        case .frameworkError(let error):
            return "Framework error: \(error.localizedDescription)"
        }
    }
}

/// Framework-specific errors
public struct FrameworkError: LocalizedError {
    public let framework: LLMFramework
    public let underlying: Error
    public let context: String?

    public init(framework: LLMFramework, underlying: Error, context: String? = nil) {
        self.framework = framework
        self.underlying = underlying
        self.context = context
    }

    public var errorDescription: String? {
        var description = "\(framework.rawValue) error: \(underlying.localizedDescription)"
        if let context = context {
            description += " (\(context))"
        }
        return description
    }
}
